ambry-tools/src/test/java/com.github.ambry/store/BlobIdTransformerTest.java

```diff
@@ -0,0 +1,388 @@
+      new BlobIdTransformer(mockStoreKeyConverter, null, null);
+      fail("Did not throw NullPointerException");
+    } catch (NullPointerException e) {
+      //expected
+    }
+  }
+
+  /**
+   * Tests using the transformer with null input to the transform method
+   * @throws Exception
+   */
+  @Test
+  public void testNullTransformInput() throws Exception {
+    try {
+      transformer.transform(null);
+      fail("Did not throw NullPointerException");
+    } catch (NullPointerException e) {
+      //expected
+    }
+  }
+
+  /**
+   * Tests using the transformer with StoreCopier.Message inputs that have null components
+   * @throws Exception
+   */
+  @Test
+  public void testNullComponentsTransformInput() throws Exception {
+    MessageInfo messageInfo =
+        new MessageInfo(createBlobId(VERSION_1_UNCONVERTED), 123, (short) 123, (short) 123, 0L);
+    InputStream inputStream = null;
+    //null inputStream
+    StoreCopier.Message message = new StoreCopier.Message(messageInfo, inputStream);
+    try {
+      transformer.transform(message);
+      fail("Did not throw NullPointerException");
+    } catch (NullPointerException e) {
+      //expected
+    }
+    //null messageInfo
+    message = new StoreCopier.Message(null, new ByteArrayInputStream(new byte[30]));
+    try {
+      transformer.transform(message);
+      fail("Did not throw NullPointerException");
+    } catch (NullPointerException e) {
+      //expected
+    }
+  }
+
+  private BlobId createBlobId(String hexBlobId) throws IOException {
+    return new BlobId(hexBlobId, clusterMap);
+  }
+
+  private void putRowInStoreKeyConverter(MockStoreKeyConverter storeKeyConverter, BlobIdStringPair blobIdStringPair) throws IOException {
+    storeKeyConverter.put(createBlobId(blobIdStringPair.getUnconverted()), createBlobId(blobIdStringPair.getConverted()));
+  }
+
+  private MockStoreKeyConverter createAndSetupMockStoreKeyConverter(BlobIdStringPair[] blobIdStringPairs) throws IOException {
+    MockStoreKeyConverter storeKeyConverter = new MockStoreKeyConverter();
+    for (BlobIdStringPair blobIdStringPair : blobIdStringPairs) {
+      putRowInStoreKeyConverter(storeKeyConverter, blobIdStringPair);
+    }
+    return storeKeyConverter;
+  }
+
+  private String notEqualPrint(int loc, byte[] actual, byte[] expect) {
+    StringBuilder sb = new StringBuilder();
+    sb.append("Not equal at byte ")
+        .append(loc)
+        .append("\n")
+        .append("Actual: ")
+        .append(Hex.encodeHexString(actual))
+        .append("\n")
+        .append("Expect: ")
+        .append(Hex.encodeHexString(expect))
+        .append("\n");
+    return sb.toString();
+  }
+
+  private void assertInputStreamEqual(InputStream actual, InputStream expect) throws IOException {
+    byte[] actualBuf = inputStreamToByteArray(actual);
+    byte[] expectBuf = inputStreamToByteArray(expect);
+    assertTrue("Lengths are not equal, actual: " + actualBuf.length + ", expect: " + expectBuf.length,
```

Comment:
you can instead use `assertArrayEquals()` that does the length check + tells you exactly which byte differed. So you can get rid of `notEqualPrint()`